### PR TITLE
Add Intel Advanced Matrix Extensions (AMX) support to ggml

### PR DESCRIPTION
**Pull request by mingfeima**
This PR improves intel server CPU performance with intel advanced matrix extension (AMX).
AMX is a new built-in accelerator for gemm starting from 4th gen Xeon:   https://www.intel.com/content/www/us/en/products/docs/accelerator-engines/advanced-matrix-extensions/overview.html

The basic idea is pretty much the same as what i have done in PyTorch https://github.com/pytorch/pytorch/pull/117475 for the int4 and int8 mixed dtype gemms.

#### Features
* now it supports Q4_0, Q4_1, Q8_0 quantized format ( I picked up these formats based on current support from `__ARM_FEATURE_MATMUL_INT8`), more support will be added in the feature. Kernels are placed in `ggml-amx.cpp` since I don't want to mess up with ggml.c which is already very complexed and the amx kernels could also be more complexed in future if add more qformat support.
* implement fast weight only quantized gemm kernels with amx, q4 will be unpacked to q8 and doing s8s8 or u8s8 gemm.
* implement a fast path when batch dimension is 1 for gemv, when batch dimension is small vnni is usually faster than amx because amx has larger overhead.
* implement double buffering for post processing (applying scales): the scales are stored in f16 in gguf and the tmul has to be interleaved with f32 instructions. So it won't hit the hardware compute limit as we can not run tmul in series. Applying double buffering here improves 8%-11% performance.
* the amx and vnni kernels have numerically identical results with current avx or avx2 kernels from `ggml-quant.c`.
* amx and kernels will be compiled automatically on CPUs with the hardware support, otherwise not.

#### Performance
* results from llama2-7b-q4_0, about **2x speed up** for the text generation. Collected on Intel (R) Xeon (R) CPU Max 9480:

1. before
```
llama_print_timings:        load time =     533.79 ms
llama_print_timings:      sample time =       7.29 ms /   200 runs   (    0.04 ms per token, 27453.67 tokens per second)
llama_print_timings: prompt eval time =      77.35 ms /     6 tokens (   12.89 ms per token,    77.57 tokens per second)
llama_print_timings:        eval time =    9333.20 ms /   199 runs   (   46.90 ms per token,    21.32 tokens per second)
llama_print_timings:       total time =    9487.99 ms /   205 tokens
```
2. after
```
llama_print_timings:        load time =     549.56 ms
llama_print_timings:      sample time =       3.73 ms /    96 runs   (    0.04 ms per token, 25751.07 tokens per second)
llama_print_timings: prompt eval time =      67.38 ms /     6 tokens (   11.23 ms per token,    89.05 tokens per second)
llama_print_timings:        eval time =    2245.79 ms /    95 runs   (   23.64 ms per token,    42.30 tokens per second)
llama_print_timings:       total time =    2346.99 ms /   101 tokens
```

* results from `benchmark-matmult` (metic: gFlops):

cores | before | after | speedup
-- | -- | -- | --
1 | 50.67 | 260.04 | 5.13
4 | 171.92 | 1026.29 | 5.97
16 | 192.38 | 2143.1 | 11.14
32 | 263.7 | 3694.85 | 14.01

TODO:
* ~add more quantized dtype support~
* ~add bf16 gemm support with amx-bf16 (using avx512-bf16 for gemv)~
* ~add f16 gemm support with amx-f16 (using avx512-f16 for gemv)~

I also noticed from vtune that some pointwise operators need additional optimization, e.g. softmax, etc. Will handle them later on.

### CODE REVIEW

**Comment by mingfeima** [comment]
This PR also adds openmp support since the original pthead sync is done via atomic which has a very high overhead on server CPUs (and the sync has to be done very frequently for each operator launch). I noticed https://github.com/ggerganov/llama.cpp/pull/7606 is also doing this, this should also work.

**Comment by airMeng** [comment]
BTW why AMX will greatly improve next token latency?

**Comment by NeoZhangJianyu** [comment]
Here is my suggestion:
1. update the README.md 
 Explain the condition that AMX will be used to speed up inference, like hardware, build parameter.
 How to run CI

2. CMake (optional)
 Most backends use CMake to build. It's great if support CMake to build it.

**Comment by mingfeima** [comment]
> BTW why AMX will greatly improve next token latency?

I also wrote an vnni kennel for gemv cases.

**Comment by mingfeima** [comment]
> Here is my suggestion:
> 
> 1. update the README.md
>    Explain the condition that AMX will be used to speed up inference, like hardware, build parameter.
>    How to run CI
> 2. CMake (optional)
>    Most backends use CMake to build. It's great if support CMake to build it.

sure, the BKMs for intel also need to be updated.

**Comment by mingfeima** [comment]
**Updates**: f16 support added

Right now this patch only has a avx512 kernel which is doing fma with `avx512f` (did not use `avx512-fp16` here as `_mm512_fmadd_ph` is doing accumulation with 16 bits). The amx kernel will be added once the 6th gen Xeon is released which has `amx-f16` support).

Also postpone bf16 amx kernels support to align with f16 amx timeline. Since bf16 is not that common is gguf.

**Performance**: Tested on `Meta-Llama-3-8B-Instruct-fp16.gguf`, about **1.6x** performance improvement in test generation, tested on Intel (R) Xeon (R) CPU Max 9480 (the 4th gen Xeon):

```
### before:
llama_print_timings:        eval time =   20867.58 ms /   199 runs   (  104.86 ms per token,     9.54 tokens per second)

### after:
llama_print_timings:        eval time =   13214.77 ms /   199 runs   (   66.41 ms per token,    15.06 tokens per second)
```

**Comment by ggerganov** [comment]
Now that we have the necessary capabilities to implement backends such as the BLAS backend (#6210), would it make sense to implement a similar AMX backend and put this implementation there. @slaren what do you think?

**Comment by slaren** [comment]
I think it would be better to leave the implementation as is instead of moving it to a different backend, the performance would be slightly better, and I don't really see a good reason to split the CPU backend into multiple backends. The changes in `ggml.c` should not be necessary now that openmp is already used by default.

**Comment by mingfeima** [comment]
> I think it would be better to leave the implementation as is instead of moving it to a different backend, the performance would be slightly better, and I don't really see a good reason to split the CPU backend into multiple backends. The changes in `ggml.c` should not be necessary now that openmp is already used by default.

AMX is a new built-in accelerator available from the 4th generation of Xeon, the Intel sever CPU, [link](https://www.intel.com/content/www/us/en/products/docs/accelerator-engines/advanced-matrix-extensions/overview.html). So this PR is actually trying to improve the performance of llama.cpp on intel server CPUs.  And AMX is not equal to the concept of BLAS.

@slaren I don't quite get your idea, should I continue with `ggml-amx.cpp` or move the optimizations to somewhere else?
My general idea is putting all the AMX related optimizations in a single file which would be easier to maintain. The current  available Xeons (the 4th gen and the 5th gen) have the same ISA, but the 6th gen Xeon has two different types: E core and P core. 6th gen Xeon will be launched very soon, so I need to update the AMX related optimizations for the new hardware in near future: adding amx-f16 kernels.

OMP changes in ggml.c shall be gone after rebasing. Currently I am working on QK_K AMX kernels and I will clear up this PR once it is done.

**Comment by slaren** [comment]
> should I continue with `ggml-amx.cpp` or move the optimizations to somewhere else?

I was responding to @ggerganov suggestion to move the implementation to a different backend similar to the BLAS backend. I think you should continue as is.

**Comment by ggerganov** [comment]
Ok, let's proceed as is

**Comment by nai-kon** [comment]
Is there any progress?
I am really looking forward to the AMX support.

**Comment by mingfeima** [comment]
> Is there any progress? I am really looking forward to the AMX support.

Recently I got distracted by some other tasks, I use my spare time to work on this project as this is not an official task from my employer. Currently I am working on the Q4K quant format, have to say that it is much more complexed...  Anyway it's about to be finished.

**Comment by mingfeima** [comment]
Added AMX and VNNI kernels for `Q4_K`, `Q5_K`, `Q6_K`, `IQ4_XS`.

File: Makefile [hunk 1]
```diff
@@ -92,11 +92,6 @@ GGML_METAL := 1
 DEPRECATE_WARNING := 1
 endif
 
-ifdef LLAMA_OPENMP
-GGML_OPENMP := 1
-DEPRECATE_WARNING := 1
-endif
-
```
Comment (ggerganov) [inline Makefile:99]: This should not be removed

File: ggml/src/ggml.c [hunk 2]
```diff
@@ -320,6 +329,11 @@ static ggml_fp16_t ggml_table_gelu_quick_f16[1 << 16];
 // precomputed f32 table for f16 (256 KB) (ggml-impl.h)
 float ggml_table_f32_f16[1 << 16];
 
+#if GGML_USE_AMX
+// global flag for amx init
+static bool ggml_amx_initialized = false;
+#endif
+
```
Comment (ggerganov) [inline ggml/src/ggml.c:336]: Should we move this flag inside the `ggml-amx` implementation?

```c
    ggml_amx_init();

    ...

    if (ggml_amx_is_init()) {
        ...
    }
```

**Review by ggerganov — COMMENTED** [review]
I like that the code is very well isolated from the rest of the codebase. Haven't reviewed the `mmq.cpp` source in details yet and it would be difficult without the appropriate hardware, but I think that's alright as we can easily determine that there won't be side effects to the rest of the codebase.

Wondering how we could add some tests for this functionality.

Overall, seems good to me. @slaren What do you think?

File: Makefile (hunk 1 above)
Comment (slaren) [reply Makefile:99]: I think it should be removed, `GGML_OPENMP` is not an option in the `Makefile`. Since OpenMP is enabled by default, the only option the user can specify is `GGML_NO_OPENMP` to disable it.

File: ggml/src/ggml.c (hunk 2 above)
Comment (mingfeima) [reply ggml/src/ggml.c:336]: ok I will give it a try, not sure if this is feasible.

**Comment by mingfeima** [comment]
> I like that the code is very well isolated from the rest of the codebase. Haven't reviewed the `mmq.cpp` source in details yet and it would be difficult without the appropriate hardware, but I think that's alright as we can easily determine that there won't be side effects to the rest of the codebase.
> 
> Wondering how we could add some tests for this functionality.
> 
> Overall, seems good to me. @slaren What do you think?

Yeah... the CI is a big problem. I will try to find some internal sponsor and then we can use our company cloud, that would be the best. Otherwise, we will have to go the emulator.

@ggerganov i was wondering how the `Ascend910B3` functionalities are tested in the CI ? Does Huawei provides the CI support ? And also how about the aarch64 functionalities for arm servers ?

**Comment by ggerganov** [comment]
We don't have CI for the CANN backend either. For aarch64, I'm planning to try to rent an Arm machine on the Azure cloud when they become [available](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/general-purpose/dpsv6-series?tabs=sizebasic) and if they are not too expensive

**Comment by NeoZhangJianyu** [comment]
I think the Sapphire Rapids Xeon (4th generation Xeon) support AMX.
In Azure, DCesv5-series and DCedsv5-series are powered by Intel® 4th Generation Xeon® Scalable processors (https://learn.microsoft.com/en-us/azure/virtual-machines/dcesv5-dcedsv5-series).
They should support AMX. 
It possible to build CI on it.

**Comment by gyusang** [comment]
Hi, I noticed some quantization issues in `mmq.cpp`.
https://github.com/mingfeima/llama.cpp/blob/74bb1eb52be7d9b9eb484d156d24a474dd09f278/ggml/src/ggml-amx/mmq.cpp#L1183-L1195
Here, we are using a single scale `vd0` for all 16x32 weights. However, Q8_0 uses scale parameter per `blck_size=32` elements.
1. Is this compensated by creating a single scale parameter for all 16x32 weights? I don't see any code in `pack_B`, etc.
2. Wouldn't this be another quantization method if we use a single scale for 512 weights? This would have different result compared with existing Q8_0 AVX-based methods.

**Comment by mingfeima** [comment]
> Hi, I noticed some quantization issues in `mmq.cpp`. https://github.com/mingfeima/llama.cpp/blob/74bb1eb52be7d9b9eb484d156d24a474dd09f278/ggml/src/ggml-amx/mmq.cpp#L1183-L1195 Here, we are using a single scale `vd0` for all 16x32 weights. However, Q8_0 uses scale parameter per `blck_size=32` elements.
> 
> 1. Is this compensated by agreeing with a single scale parameter for all 16x32 weights? I don't see any code in `pack_B`, etc doing so.
> 2. Wouldn't this be another quantization method if we use a single scale for 512 weights? This would have different result compared with existing Q8_0 AVX-based methods.

the weight packing for Q8_0 is here https://github.com/mingfeima/llama.cpp/blob/74bb1eb52be7d9b9eb484d156d24a474dd09f278/ggml/src/ggml-amx/mmq.cpp#L866-L873

each weight block of 16x32 (NxK) is stored in the format of (KxN) so that we can do FMA here, and this block will have 16 scales (d0), it is packed as a contiguous vector of 1x16, the dtype is f16. So to sum up, the scale is a 256-bit vector which corresponds to 16 columns. So it is not a "single scale parameter for all 16x32 weights". If the computation is wrong, the llm will talk like crazy.

**Comment by mingfeima** [comment]
@ggerganov On Azure, `DCesv5` and `ECesv5` instances have intel AMX support, they are all 4th gen Xeon (codename Sapphire Rapids):
https://azure.microsoft.com/en-us/updates/confidential-vms-with-intel-tdx-dcesv5-ecesv5/
https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/general-purpose/dcesv5-series?tabs=sizebasic

Is that possible to use those instances for CI ?
The amx features will be compiled on CPUs with AMX support by default, the requirement for minimum gcc version is gcc-11.

**Comment by ggerganov** [comment]
Thanks for letting me know - I just added an AMX VM (EC8eds v5) to the `ggml-ci` fleet:

2024-08-08 10 24 49 github com 0f97b4e43c09

https://github.com/ggml-org/ci/tree/results/llama.cpp/15/fa07a5c564d3ed7e7eb64b73272cedb27e73ec/ggml-5-x86-amx-cc#summary

It won't run on this PR since [ggml-ci](https://github.com/ggerganov/llama.cpp/tree/master/ci) runs only on branches in this repository. So the AMX CI will run after we merge the PR in `master`.

I've also sent you a collaborator invite, if you'd like you will be able to push branches in this repository and be able to run the CI prior to merging in the future.

File: ggml/src/ggml.c (hunk 2 above)
Comment (mingfeima) [reply ggml/src/ggml.c:415]: changed in https://github.com/ggerganov/llama.cpp/pull/8998
i moved `ggml_amx_init()` from ggml.c to ggml-amx/mmq.cpp, it will be called for the 1st time launch.

**Comment by artun42** [comment]
Hi, just checking in - any progress to merging this to master?

**Comment by ggerganov** [comment]
It's already merged: https://github.com/ggml-org/llama.cpp/pull/8998

**Comment by mtcl** [comment]
If I have a Sapphire Rapids processor which is AMX enabled,  how do i ensure that I have them enabled in llama.cpp?

currently I am building it with 

```bash
cmake -B build -DGGML_CUDA=ON -DGGML_RPC=ON
cmake --build build --config Release -j 56
```

should i use argument like this to add at the end?
```bash
-DGGML_USE_AMX=ON
```
